pkg/remotemetrics/resources.go

```diff
@@ -0,0 +1,140 @@
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/liqotech/liqo/pkg/consts"
+	"github.com/liqotech/liqo/pkg/virtualKubelet/forge"
+)
+
+type resourceGetter struct {
+	cl client.Client
+}
+
+// NewResourceGetter creates a new resource getter.
+func NewResourceGetter(cl client.Client) ResourceGetter {
+	return &resourceGetter{
+		cl: cl,
+	}
+}
+
+// GetNamespaces returns the names of all namespaces in the cluster owned by the remote clusterID.
+func (m *resourceGetter) GetNamespaces(ctx context.Context, clusterID string) []MappedNamespace {
+	namespaces := &corev1.NamespaceList{}
+
+	err := m.cl.List(ctx, namespaces)
+	utilruntime.Must(err)
+
+	res := []MappedNamespace{}
+	for i := range namespaces.Items {
+		namespace := &namespaces.Items[i]
+
+		if getClusterIDFromNamespace(namespace) != clusterID {
+			continue
+		}
+		if namespace.Annotations == nil || namespace.Annotations[consts.RemoteNamespaceOriginalNameAnnotationKey] == "" {
+			continue
+		}
+
+		res = append(res, MappedNamespace{
+			Namespace:    namespace.GetName(),
+			OriginalName: namespace.Annotations[consts.RemoteNamespaceOriginalNameAnnotationKey],
+		})
+	}
+
+	klog.V(2).Infof("Scraping namespaces %+v for cluster id %s", res, clusterID)
+	return res
+}
+
+// GetPods returns the names of all pods in the cluster owned by the remote clusterID and scheduled in the given node.
+func (m *resourceGetter) GetPods(ctx context.Context, clusterID, node string) []string {
+	pods := &corev1.PodList{}
+
+	err := m.cl.List(ctx, pods)
+	utilruntime.Must(err)
+
+	res := []string{}
+	for i := range pods.Items {
+		pod := &pods.Items[i]
+
+		if getClusterIDFromPod(pod) == clusterID && pod.Spec.NodeName == node {
+			res = append(res, pod.Name)
+		}
+	}
+
+	klog.V(2).Infof("Scraping pods %+v for cluster id %s and node %s", res, clusterID, node)
+	return res
+}
+
+// GetNodes returns the names of all physical nodes in the cluster.
+func (m *resourceGetter) GetNodes(ctx context.Context) []string {
+	nodes := &corev1.NodeList{}
+
+	// we exclude virtual nodes to avoid infinite loops, both for bidirectional peerings
+	// and for cycling peerings A -> B -> C -> A that would lead to non-terminating HTTP requests.
+	realNode, err := labels.NewRequirement(consts.TypeLabel, selection.NotIn, []string{consts.TypeNode})
+	utilruntime.Must(err)
+
+	err = m.cl.List(ctx, nodes, client.MatchingLabelsSelector{
+		Selector: labels.NewSelector().Add(*realNode),
+	})
+	utilruntime.Must(err)
+
+	res := make([]string, len(nodes.Items))
+	for i := range nodes.Items {
+		node := &nodes.Items[i]
+		res[i] = node.Name
+	}
+
+	klog.V(2).Infof("Scraping nodes %+v", res)
+	return res
+}
+
+func getClusterIDFromNamespace(namespace *corev1.Namespace) string {
+	if namespace.Annotations == nil {
+		return ""
+	}
+
+	var clusterID string
+	var ok bool
+	if clusterID, ok = namespace.Annotations[consts.RemoteClusterID]; !ok || clusterID == "" {
+		return ""
+	}
```

Comment:
Could avoid defining the variables in advance by changing the if logic (same below)